jumanji/environments/swarms/common/updates.py

```diff
@@ -0,0 +1,153 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+import chex
+import esquilax
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.swarms.common import types
+
+
+@esquilax.transforms.amap
+def update_velocity(
+    _: chex.PRNGKey,
+    params: types.AgentParams,
+    x: Tuple[chex.Array, types.AgentState],
+) -> Tuple[float, float]:
+    """
+    Get the updated agent heading and speeds from actions
+
+    Args:
+        _: Dummy JAX random key.
+        params: Agent parameters.
+        x: Agent rotation and acceleration actions.
+
+    Returns:
+        float: New agent heading.
+        float: New agent speed.
+    """
+    actions, boid = x
+    rotation = actions[0] * params.max_rotate * jnp.pi
+    acceleration = actions[1] * params.max_accelerate
+
+    new_heading = (boid.heading + rotation) % (2 * jnp.pi)
+    new_speeds = jnp.clip(
+        boid.speed + acceleration,
+        min=params.min_speed,
+        max=params.max_speed,
+    )
+
+    return new_heading, new_speeds
+
+
+def move(pos: chex.Array, heading: chex.Array, speed: chex.Array) -> chex.Array:
+    """
+    Get updated agent positions from current speed and heading
+
+    Args:
+        pos: Agent position
+        heading: Agent heading (angle).
+        speed: Agent speed
+
+    Returns:
+        jax array (float32): Updated agent position
+    """
+    d_pos = jnp.array([speed * jnp.cos(heading), speed * jnp.sin(heading)])
+    return (pos + d_pos) % 1.0
+
+
+def update_state(
+    key: chex.PRNGKey,
+    params: types.AgentParams,
+    state: types.AgentState,
+    actions: chex.Array,
+) -> types.AgentState:
+    """
+    Update the state of a group of agents from a sample of actions
+
+    Args:
+        key: Dummy JAX random key.
+        params: Agent parameters.
+        state: Current agent states.
+        actions: Agent actions, i.e. a 2D array of action for each agent.
+
+    Returns:
+        AgentState: Updated state of the agents after applying steering
+            actions and updating positions.
+    """
+    actions = jnp.clip(actions, min=-1.0, max=1.0)
+    headings, speeds = update_velocity(key, params, (actions, state))
+    positions = jax.vmap(move)(state.pos, headings, speeds)
+
+    return types.AgentState(
+        pos=positions,
+        speed=speeds,
+        heading=headings,
+    )
+
+
+def view(
+    _key: chex.PRNGKey,
+    params: Tuple[float, float],
```

Comment:
This is a design choice in Esquilax, how broadcasting/mapping of arguments is decided is done positionally (so in this case the params are broadcast rather than mapped across the mapped interactions). 
It could be nice to have some alternate way to mark how arguments are to be used, but not settled on anything clean yet.